aiomultiprocess/core.py

```diff
@@ -466,3 +469,166 @@ def terminate(self) -> None:
+        self,
+        task_id: TaskID,
+        func: Callable[..., Awaitable[R]],
+        args: Sequence[Any],
+        kwargs: Dict[str, Any],
+    ) -> QueueID:
+        return choice(self.qids)
+
+    def task_done(self, _task_id: TaskID) -> None:
+        pass
+
+
+class FailedToStartProcessError(Exception):
+    pass
+
+
+class BrokenShardedPoolException(Exception):
+    pass
+
+
+class ShardedPool(Pool):
+    """
+    Execute coroutines on a pool of child processes,
+    where each having a dedicated queue, support custom scheduling.
+    """
+
+    def __init__(
+        self,
+        scheduler: ShardedPoolSchedulerBase,
+        processes: int = None,
+        initializer: Callable[..., None] = None,
+        initargs: Sequence[Any] = (),
+        maxtasksperchild: int = MAX_TASKS_PER_CHILD,
+        childconcurrency: int = CHILD_CONCURRENCY,
+    ) -> None:
+        # `loop` is scheduled by `ensure_future`, but we haven't given up control,
+        # spawn processes before `loop` runs, so tx_queues will be populated,
+        # `queue_work` will fail if `tx_queues` is empty
+        super().__init__(
+            processes, initializer, initargs, maxtasksperchild, childconcurrency
+        )
+        del self.tx_queue
+
+        self.scheduler = scheduler
+        self.tx_queues: List[multiprocessing.Queue] = []
+        self.pid_to_qid: Dict[int, QueueID] = {}
+        self._populate_processes()
+
+    async def loop(self) -> None:
+        """Maintain the pool of workers while open."""
+        while self.processes or self.running:
+            # clean up workers that reached TTL, new processes will reuse queues
+            outstanding_queues: List[QueueID] = []
+            for process in self.processes:
+                if not process.is_alive():
+                    if process.pid not in self.pid_to_qid:
+                        raise BrokenShardedPoolException()
+                    outstanding_queues.append(self.pid_to_qid.pop(process.pid))
+                    self.processes.remove(process)
+
+            # start new workers when slots are unfilled
+            self._populate_processes(outstanding_queues)
+
+            # pull results into a shared dictionary for later retrieval
+            while True:
+                try:
+                    task_id, value, tb = self.rx_queue.get_nowait()
+                    self._results[task_id] = value, tb
+                    self.scheduler.task_done(task_id)
+
+                except queue.Empty:
+                    break
+
+            # let someone else do some work for once
+            await asyncio.sleep(0.005)
+
+    def _populate_processes(self, outstanding_queues: Optional[List[QueueID]] = None):
+        while self.running and len(self.processes) < self.process_count:
+            if outstanding_queues:
+                qid = outstanding_queues.pop()
+                tx_queue = self.tx_queues[qid]
+            else:
+                qid = QueueID(len(self.tx_queues))
```

Comment:
Is there ever a case where queues would need to be recreated?  Would it be worth it to track and increment a queue id separately to ensure IDs never get reused, similar to task IDs?

Reply:
Can totally switch to a dictionary implementation.

There should really be no deletion of queues though, as the queue is per process and there is no way to dynamically change the pool size. So actually the queues should be created in __init__ and never got deleted, and actually there should be no more queues creating afterwards. Maybe a clearer implementation is to create queues within `__init__`, then have the `qid` added to object member `outstanding_queues`, the responsibility of `_populate_processes` would then become cleaner to just assign queues to processes, not creating new queues, it is also clear that queues are only created in `__init__` so recreation would not happen.